### PR TITLE
DD4Hep: Migrate Run 1 Tracker Algorithms

### PR DESCRIPTION
**Pull request by ianna**
* Move DD CMS example under our control for further development
* Move Run 1 Tracker algorithms in CMSSW

### CODE REVIEW

**Comment by ianna** [comment]
@smuzaffar - FYI:

```
$ cmsRun DetectorDescription/DDCMS/python/dump.py
----- Begin Fatal Exception 16-Mar-2018 14:45:49 CET-----------------------
An exception of category 'MultiplePlugins' occurred while
   [0] Constructing the EventProcessor
Exception Message:
The plugin 'DDCMSDetector' is found in multiple files 
 '"pluginDetectorDescriptionDD4HepPlugins.so"'
 '"pluginDetectorDescriptionDDCMSPlugins.so"'
in directory '/build/yana/DD4HEP/CMSSW_10_1_X_2018-03-15-2300/lib/slc6_amd64_gcc630'.
The code must be changed so the plugin only appears in one plugin file. You will need to remove the macro which registers the plugin so it only appears in one of these files.
  If none of these files register such a plugin, then the problem originates in a library to which all these files link.
The plugin registration must be removed from that library since plugins are not allowed in regular libraries.
----- End Fatal Exception -------------------------------------------------

```

**Comment by ianna** [comment]
hold

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
unhold

File: DetectorDescription/DDCMS/BuildFile.xml [hunk 1]
```diff
@@ -1,2 +1,7 @@
 <use   name="FWCore/PluginManager"/>
 <use   name="dd4hep"/>
+<use   name="rootmath"/>
```
Comment (VinInn) [inline DetectorDescription/DDCMS/BuildFile.xml]: why?

File: DetectorDescription/DDCMS/interface/DDCMSPlugins.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+// DDCMS is a detector description convention developed by the CMS experiment.
+//
+//==========================================================================
+#ifndef DD4HEP_DDCMS_DDCMSPLUGINS_H
+#define DD4HEP_DDCMS_DDCMSPLUGINS_H
+
+// Framework includes
+#include "DetectorDescription/DDCMS/interface/DDCMS.h"
+#include "DD4hep/Plugins.h"
+#include "CLHEP/Units/SystemOfUnits.h"
```
Comment (VinInn) [inline DetectorDescription/DDCMS/interface/DDCMSPlugins.h]: can we get rid of this?
I though we decided to obsolete CLHEP. No new code should use it anymore!

File: DetectorDescription/DDCMS/plugins/DDTrackerPhiAltAlgo.cc [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    double dphi;
+    if (number == 1 || fabs(rangeAngle-360.0*CLHEP::deg)<0.001*CLHEP::deg) 
+      dphi = rangeAngle/number;
+    else
+      dphi = rangeAngle/(number-1);
+    int copyNo = startCopyNo;
+
+    for (int i=0; i<number; i++) {
+      double phi  = startAngle + i*dphi;
+      double phix = phi - tilt + 90.*CLHEP::deg;
```
Comment (VinInn) [inline DetectorDescription/DDCMS/plugins/DDTrackerPhiAltAlgo.cc]: 90.*CLHEP::deg;  this is simply ridiculous...
please use angles in radiant

Comment (ianna) [reply DetectorDescription/DDCMS/plugins/DDTrackerPhiAltAlgo.cc]: @VinInn - I agree, but it's one to one port of existing original Tracker code:
https://cmssdt.cern.ch/lxr/source/Geometry/TrackerCommonData/plugins/DDTrackerPhiAltAlgo.cc

**Comment by smuzaffar** [comment]
please test
new dd4hep plugin build rules are available in 10.1.X IBs.

File: DetectorDescription/DDCMS/plugins/DDTrackerPhiAltAlgo.cc (hunk 3 above)
Comment (VinInn) [reply DetectorDescription/DDCMS/plugins/DDTrackerPhiAltAlgo.cc]: I see.
Then let's schedule it for a migration when some personpower will be available

Comment (ianna) [reply DetectorDescription/DDCMS/plugins/DDTrackerPhiAltAlgo.cc]: @VinInn - Yes, the right order is to improve current code, validate, then migrate, then validate again. The first can go in parallel when you'll find the personpower. You (e.g. Tracker group) will certainly keep the ownership of the code in its current location and migrate it to DD4Hep later, replacing the algorithms in this package.
This package is a merely playground for DD4Hep evaluation.

**Comment by ianna** [comment]
+1

**Comment by smuzaffar** [comment]
@ianna , I see some thread related error/warning 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22643/26957/llvm-analysis/
and I also see clang crash here https://cmssdt.cern.ch/jenkins/job/ib-any-integration/26957/consoleFull
```
17:45:40 /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-03-19-1100/src/DetectorDescription/DDCMS/plugins/DDCMSDetElementCreator.cc:339:1: warning: Non-const variable 'struct __typeFactory___dict_339 s____typeFactory___dict_339' is static and might be thread-unsafe
17:45:40 DECLARE_DD4HEP_CONSTRUCTOR(DDCMS_DetElementCreator,create_object)
17:45:40 ^
17:45:40 /cvmfs/cms-ib.cern.ch/nweek-02516/slc6_amd64_gcc630/external/dd4hep/v01-06x-fjcbno/include/DD4hep/Factories.h:273:5: note: expanded from macro 'DECLARE_DD4HEP_CONSTRUCTOR'
17:45:40     DD4HEP_PLUGINSVC_FACTORY(name,name,void*(dd4hep::Detector*,int,char**),__LINE__) }
17:45:40     ^
17:45:40 /cvmfs/cms-ib.cern.ch/nweek-02516/slc6_amd64_gcc630/external/dd4hep/v01-06x-fjcbno/include/DD4hep/Plugins.h:169:5: note: expanded from macro 'DD4HEP_PLUGINSVC_FACTORY'
17:45:40     struct DD4HEP_PLUGINSVC_CNAME(__typeFactory__,serial)  {            \
17:45:40     ^
17:45:40 1 warning generated.
17:45:41 >> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-03-19-1100/src/DetectorDescription/DDCMS/plugins/DDTrackerPhiAltAlgo.cc 
17:45:43 >> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-03-19-1100/src/DetectorDescription/DDCMS/plugins/DDTIDModulePosAlgo.cc 
17:45:45 /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-03-19-1100/src/DetectorDescription/DDCMS/plugins/DDTECPhiAlgo.cc:57:1: warning: Non-const variable 'struct __typeFactory___dict_57 s____typeFactory___dict_57' is static and might be thread-unsafe
17:45:45 DECLARE_DDCMS_DETELEMENT(DDCMS_track_DDTECPhiAlgo,algorithm)
17:45:45 ^
17:45:45 /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-03-19-1100/src/DetectorDescription/DDCMS/interface/DDCMSPlugins.h:64:5: note: expanded from macro 'DECLARE_DDCMS_DETELEMENT'
17:45:45     DD4HEP_PLUGINSVC_FACTORY(ddcms_det_element_##name,name,             \
17:45:45     ^
17:45:45 /cvmfs/cms-ib.cern.ch/nweek-02516/slc6_amd64_gcc630/external/dd4hep/v01-06x-fjcbno/include/DD4hep/Plugins.h:169:5: note: expanded from macro 'DD4HEP_PLUGINSVC_FACTORY'
17:45:45     struct DD4HEP_PLUGINSVC_CNAME(__typeFactory__,serial)  {            \
17:45:45     ^
17:45:45 1 warning generated.
17:45:45 *** glibc detected *** /cvmfs/cms-ib.cern.ch/nweek-02516/slc6_amd64_gcc630/external/llvm/5.0.0-fjcbno/bin/clang++: corrupted double-linked list: 0x0000000001d7a2a0 ***
17:45:45 ======= Backtrace: =========
17:45:45 /lib64/libc.so.6[0x31d1675dee]
17:45:45 /lib64/libc.so.6[0x31d1678ff6]
17:45:45 /cvmfs/cms-ib.cern.ch/nweek-02516/slc6_amd64_gcc630/external/llvm/5.0.0-fjcbno/bin/../lib64/libLLVMSupport.so.5(llvm_regfree+0x42)[0x7fc4ebbbf2a2]
17:45:45 /cvmfs/cms-ib.cern.ch/nweek-02516/slc6_amd64_gcc630/external/llvm/5.0.0-fjcbno/bin/../lib64/libLLVMSupport.so.5(_ZN4llvm5RegexD1Ev+0x11)[0x7fc4ebb702f1]
17:45:45 /lib64/libc.so.6(exit+0xe2)[0x31d1635992]
17:45:45 /lib64/libc.so.6(__libc_start_main+0x104)[0x31d161ed24]
17:45:45 /cvmfs/cms-ib.cern.ch/nweek-02516/slc6_amd64_gcc630/external/llvm/5.0.0-fjcbno/bin/clang++[0x40d561]
```

**Comment by ianna** [comment]
@smuzaffar - thanks! I've submitted an issue. Let's hope it will be fixed in the next iteration. Shall I close the PR?

**Pull request closed without merging**